File: leetcode/5.longest-palindromic-substring.py
# @leet start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        best_left = 0
        best_right = 0

        for i in range(1, n):
            for left, right in ((i - 1, i), (i, i)):
                while left >= 0 and right < n and s[left] == s[right]:
                    left -= 1
                    right += 1

                left += 1
                right -= 1

                if right - left > best_right - best_left:
                    best_left = left
                    best_right = right

        return s[best_left : best_right + 1]


# @leet end

File: leetcode/5.longest-palindromic-substring.ts
// @leet start
function longestPalindrome(s: string): string {
  const dp: boolean[][] = Array.from(s, () => Array.from(s, () => true));

  let maxLength = 1;
  let longest = s[0];

  for (let n = 2; n <= s.length; n++) {
    for (let i = 0; i <= s.length - n; i++) {
      const j = i + n - 1;
      dp[i][j] = s[i] === s[j] && dp[i + 1][j - 1];

      if (dp[i][j] && j - i + 1 > maxLength) {
        maxLength = j - i + 1;
        longest = s.slice(i, j + 1);
      }
    }
  }

  return longest;
}
// @leet end
